File: Parent/Parent/Airwolf/Canvas/Assignments.swift
//
// This file is part of Canvas.
// Copyright (C) 2016-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

import Foundation

import ReactiveSwift
import Marshal
import CanvasCore

extension Assignment {

    public static func getAssignment(_ session: Session, courseID: String, assignmentID: String) throws -> SignalProducer<JSONObject, NSError> {
        let request = try session.GET("/api/v1/courses/\(courseID)/assignments/\(assignmentID)", parameters: Assignment.parameters)
        return session.JSONSignalProducer(request)
    }

    public static func refresher(_ session: Session, studentID: String, courseID: String, assignmentID: String) throws -> Refresher {
        let remote = try Assignment.getAssignment(session, courseID: courseID, assignmentID: assignmentID).map { [$0] }.map(insert(studentID, forKey: "studentID"))

        let local = predicate(courseID, assignmentID: assignmentID)
        let context = try session.assignmentsManagedObjectContext(studentID)
        let sync = Assignment.syncSignalProducer(local, inContext: context, fetchRemote: remote)

        let key = cacheKey(context, [studentID, courseID, assignmentID])
        return SignalProducerRefresher(refreshSignalProducer: sync, scope: session.refreshScope, cacheKey: key)
    }

    public static func observer(_ session: Session, studentID: String, courseID: String, assignmentID: String) throws -> ManagedObjectObserver<Assignment> {
        let pred = predicate(courseID, assignmentID: assignmentID)
        let context = try session.assignmentsManagedObjectContext(studentID)
        return try ManagedObjectObserver<Assignment>(predicate: pred, inContext: context)
    }
}

File: rn/Teacher/src/canvas-api/apis/assignments.js
//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import httpClient from '../httpClient'
import cloneDeep from 'lodash/cloneDeep'
import { paginate, exhaust } from '../utils/pagination'

export function getAssignments (courseID: string, include: string[] = []): ApiPromise<Assignment[]> {
  const url = `courses/${courseID}/assignments`
  const options = { params: { include } }
  return exhaust(paginate(url, options))
}

export function getAssignment (courseID: string, assignmentID: string): ApiPromise<Assignment> {
  const url = `courses/${courseID}/assignments/${assignmentID}`
  const options = {
    params: {
      include: ['overrides', 'observed_users'],
      all_dates: 'true',
    },
  }

  return httpClient.get(url, options)
}

export function updateAssignment (courseID: string, assignment: Assignment): ApiPromise<Assignment> {
  // For whatever reason, overrides has a different name when you try to update it
  const updatedAssignment = cloneDeep(assignment)
  if (updatedAssignment.overrides) {
    updatedAssignment.assignment_overrides = updatedAssignment.overrides
    delete updatedAssignment.overrides
  }

  const url = `courses/${courseID}/assignments/${assignment.id}`
  return httpClient.put(url, {
    assignment: updatedAssignment,
  })
}

export function getAssignmentGradeableStudents (courseID: string, assignmentID: string): ApiPromise<UserDisplay[]> {
  const url = `courses/${courseID}/assignments/${assignmentID}/gradeable_students`

  let options = {
    params: {
      per_page: 100,
    },
  }
  const students = paginate(url, options)
  return exhaust(students)
}
